ta_services/secure_key_services/src/handle.c

```diff
@@ -62,22 +67,43 @@ uint32_t handle_get(struct handle_db *db, void *ptr)
 	return n;
 }
 
+/*
+ * Protect from speculative access based on out bound handle values:
+ * Generate a validation mask from signed difference from max valid handle
+ * value to argument handle value:
+ * - out bound handle values generate a null mask.
+ * - handle values that can be seen as negative index generate a null mask.
+ * This scheme expects value 0 is an invalid handle value and should
+ * always return an invalid reference: here always a NULL reference.
+ */
+static size_t guard_handle(struct handle_db *db, uint32_t handle)
+{
+	uint32_t mask_neg = ((int32_t)handle) >> (sizeof(handle) * 8 - 1);
+	uint32_t udiff = db->max_ptrs - handle;
+	int32_t diff = ((int32_t)udiff) >> (sizeof(udiff) * 8 - 1);
+	uint32_t mask_ovf = diff;
+
+	return handle & ~mask_ovf & ~mask_neg;
```

Comment:
I'm worried that the compiler might still use "forbidden" (conditional) instructions. See the commit message in https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?h=v4.17-rc1&id=022620eed3d0bc4bf2027326f599f5ad71c2ea3f.